### ISSUE TITLE
StackOverflowError in DefaultBuilderProvider when a type has a public static generic method that returns the type

### ISSUE THREAD

**Issue by remipassmoilesel** [mapstruct/mapstruct#1751]
Hi !
I am playing a little with vavr (http://www.vavr.io) and mapstruct, but if I try to map beans with vavr lists, it fails with a stackoverflow error, even if I provide an implementation.

Mapper: 
```java
@Mapper
public interface PersonMapper {

    @Mappings(
            @Mapping(source = "age", target = "ageOfPerson")
    )
    PersonEntity toEntity(PersonDto dto);

    java.util.List<PersonEntity> toEntities(java.util.List<PersonDto> dtos);

    // Cause a stackoverlow. Why ?
    // List<PersonEntity> toVavrEntityList(List<PersonDto> dtos);

    // The same !

    // default List<PersonEntity> toVavrEntityList(List<PersonDto> dtos){
    //     return List.ofAll(toEntities(dtos.asJava()));
    // }

    @Mappings(
            @Mapping(source = "ageOfPerson", target = "age")
    )
    PersonDto toDto(PersonEntity entity);

    java.util.List<PersonDto> toDtos(java.util.List<PersonEntity> entities);

    default java.util.List<PersonDto> toDtosWithFancyRules(java.util.List<PersonEntity> entities) {
        return toDtos(entities).stream()
                .map(person -> person.toBuilder().age(person.getAge() * 2).build())
                .collect(Collectors.toList());
    }

}

```

Error:
```java
.../lightplayground/mapstruct/PersonMapper.java:11: error: Internal error in the mapping processor: java.la.StackOverflowError   
     at jdk.compiler/com.sun.tools.javac.util.Name.toString(Name.java:152)
     at jdk.compiler/com.sun.tools.javac.util.Name.charAt(Name.java:66) 
     at java.base/java.util.regex.Pattern$Slice.match(Pattern.java:4082)
     at java.base/java.util.regex.Pattern$Begin.match(Pattern.java:3656)
     at java.base/java.util.regex.Matcher.match(Matcher.java:1755)
     at java.base/java.util.regex.Matcher.matches(Matcher.java:713)     
     at org.mapstruct.ap.spi.DefaultBuilderProvider.shouldIgnore(DefaultBuilderProvider.java:281)
     at org.mapstruct.ap.spi.DefaultBuilderProvider.findBuildMethods(DefaultBuilderProvider.java:229)   
     at org.mapstruct.ap.spi.DefaultBuilderProvider.findBuilderInfo(DefaultBuilderProvider.java:167)    
     at org.mapstruct.ap.spi.DefaultBuilderProvider.findBuilderInfo(DefaultBuilderProvider.java:96)     
     at org.mapstruct.ap.internal.model.common.TypeFactory.findBuilder(TypeFactory.java:530)    
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:189)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:180)
     at org.mapstruct.ap.internal.model.common.BuilderType.create(BuilderType.java:98)  
     at org.mapstruct.ap.internal.model.common.Type.<init>(Type.java:160)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:280)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:180)
     at org.mapstruct.ap.internal.model.common.BuilderType.create(BuilderType.java:98)  
     at org.mapstruct.ap.internal.model.common.Type.<init>(Type.java:160)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:280)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:180)
     at org.mapstruct.ap.internal.model.common.BuilderType.create(BuilderType.java:98)  
     at org.mapstruct.ap.internal.model.common.Type.<init>(Type.java:160)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:280)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:180)
     at org.mapstruct.ap.internal.model.common.BuilderType.create(BuilderType.java:98)  
     at org.mapstruct.ap.internal.model.common.Type.<init>(Type.java:160)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:280)
...
```

And beans are very simples. You can find a working example online:

Repo: 
https://github.com/remipassmoilesel/mapstruct-issue-03-2019

Mapper:
https://github.com/remipassmoilesel/mapstruct-issue-03-2019/blob/master/src/main/java/org/remipassmoilesel/lightplayground/mapstruct/PersonMapper.java

Beans:
https://github.com/remipassmoilesel/mapstruct-issue-03-2019/tree/master/src/main/java/org/remipassmoilesel/lightplayground/mapstruct

Basic test:
https://github.com/remipassmoilesel/mapstruct-issue-03-2019/blob/master/src/test/java/org/remipassmoilesel/lightplayground/mapstruct/MapstructTest.java

Thanks for your help, and this awesome work on MapStruct !

**Comment by remipassmoilesel** [comment]
Any idea ?

**Comment by Decat-SimonA** [comment]
It looks like a regression in `1.3.0.Final`, if you rollback to `1.2.0.Final`

```java
  default List<PersonEntity> toVavrEntityList(List<PersonDto> dtos) {
    return List.ofAll(toEntities(dtos.asJava()));
  }
```
 works fine

BTW: this is simpler
```java
  default List<PersonEntity> toVavrEntityList(List<PersonDto> dtos) {
    return dtos.map(this::toEntity);
  }
```

**Comment by remipassmoilesel** [comment]
Hi @Decat-SimonA  ! Thanks for your help, I will test it soon.

**Comment by filiphr** [comment]
Thanks for reporting this @remipassmoilesel. The reason why this is happening is due to the added builder support and MapStruct trying to find a builder for the vavr list. The default implementation should work. 

Mapping into a vavr list was never supported, I am not sure whether it is worth adding such support considering how easy it looks to convert that list with a single default method. If more complex vavr support is needed then I think a new issue is needed in order to see how the community feels about it and maybe someone from the community can jump on it and add support :).

I'll keep this issue open until we fix the bug with the stackoverflow

**Comment by filiphr** [comment]
@remipassmoilesel this would be fixed with PR #1807. Is it possible to try the PR on your codebase? 

The default implementation with the Vavr list should now work. 

I'll rename the issue to match the actual cause of the problem

**Comment by remipassmoilesel** [comment]
Hi @filiphr, thanks for your help. I just tried master branch this morning, and it does not work, either with default implementation or not.

This repository can show you the issue: https://github.com/remipassmoilesel/mapstruct-issue-03-2019
(you will have to build mapstruct 1.4 SNAPSHOT locally before)

```
remipassmoilesel@XXXXXXXX: ~/projects/mapstruct-issue master   
$ 😝 ./gradlew clean test 

> Task :compileJava FAILED                                                                                                                                                                                         /home/remipassmoilesel/projects/mapstruct-issue/src/main/java/org/remipassmoilesel/lightplayground/mapstruct/PersonMapper.java:11: error: Internal error in the mapping processor: java.lang.StackOverflowError         at com.sun.tools.javac.code.Types$18.visitClassType(Types.java:2318)    at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:786)        at com.sun.tools.javac.code.Types$UnaryVisitor.visit(Types.java:4652)   at com.sun.tools.javac.code.Types.supertype(Types.java:2315)    at com.sun.tools.javac.code.Types$13.visitClassType(Types.java:1966)    at com.sun.tools.javac.code.Types$13.visitClassType(Types.java:1955)    at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:786)        at com.sun.tools.javac.code.Types$DefaultTypeVisitor.visit(Types.java:4571)     at com.sun.tools.javac.code.Types.asSuper(Types.java:1952)      at com.sun.tools.javac.code.Types$4.visitClassType(Types.java:921)      at com.sun.tools.javac.code.Types$4.visitClassType(Types.java:844)      at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:786)        at com.sun.tools.javac.code.Types$DefaultTypeVisitor.visit(Types.java:4571)     at com.sun.tools.javac.code.Types.isSubtype(Types.java:840)     at com.sun.tools.javac.code.Types.isSubtype(Types.java:804)     at com.sun.tools.javac.code.Ty
....
```

**Comment by stephanpelikan** [comment]
Maybe the issue #1716 is a similar aspect of the same problem caused by another situation?

**Comment by filiphr** [comment]
@remipassmoilesel sorry for my late reply. What I meant is to try my branch (https://github.com/filiphr/mapstruct/tree/1751), it has not been merged to master yet 😄.

@stephanpelikan I think that indeed #1716 has the same problem.

**Comment by filiphr** [comment]
We've merged the PR with master, so you can try again with master.

**Comment by remipassmoilesel** [comment]
Hi @filiphr, new issue this morning:

```
$ cd mapstruct
$ git log
* 9c33199a - (HEAD -> master, origin/master, origin/HEAD) Improve terms in qualifier docs (#1814) (il y a 51 minutes) <Matt Drees>
* 33710584 - #1742 & #1661 refactoring and making builder optional (#1811) (il y a 13 heures) <Sjaak Derksen>
* 60c159a0 - #1751 Fix handling of possible builder creation methods with generic signature (il y a 13 heures) <Filip Hrisafov>

$ cd mapstruct-issue # remote: git@github.com:remipassmoilesel/mapstruct-issue-03-2019.git
$ ./gradle test

/home/remipassmoilesel/projects/mapstruct-issue/src/main/java/org/remipassmoilesel/lightplayground/mapstruct/PersonMapper.java:11: error: Internal error in the mapping processor: java.util.ServiceConfigurationError: org.mapstruct.ap.internal.processor.ModelElementProcessor: Error reading configuration file     at java.util.ServiceLoader.fail(ServiceLoader.java:232)         at java.util.ServiceLoader.parse(ServiceLoader.java:309)        at java.util.ServiceLoader.access$200(ServiceLoader.java:185)   at java.util.ServiceLoader$LazyIterator.hasNextService(ServiceLoader.java:357)          at java.util.ServiceLoader$LazyIterator.hasNext(ServiceLoader.java:393)         at java.util.ServiceLoader$1.hasNext(ServiceLoader.java:474)    at org.mapstruct.ap.MappingProcessor.getProcessors(MappingProcessor.java:314)   at org.mapstruct.ap.MappingProcessor.processMapperTypeElement(MappingProcessor.java:271)        at org.mapstruct.ap.MappingProcessor.processMapperElements(MappingProcessor.java:226)   at org.mapstruct.ap.MappingProcessor.process(MappingProcessor.java:161)         at org.gradle.api.internal.tasks.compile.processing.DelegatingProcessor.process(DelegatingProcessor.java:62)    at org.gradle.api.internal.tasks.compile.processing.NonIncrementalProcessor.process(NonIncrementalProcessor.java:45)    at org.gradle.api.internal.tasks.compile.processing.DelegatingProcessor.process(DelegatingProcessor.java:62)    at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor.access$401(TimeTrackingProcessor.java:37)     at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor$5.create(TimeTrackingProcessor.java:99)       at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor$5.create(TimeTrackingProcessor.java:96)       at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor.track(TimeTrackingProcessor.java:117)
     at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor.process(TimeTrackingProcessor.java:96)        at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)         at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:705)   at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)    at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1035)    at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1176)         at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)     at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)         at com.sun.tools.javac.main.Main.compile(Main.java:523)         at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)         at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)   at org.gradle.api.internal.tasks.compile.AnnotationProcessingCompileTask.call(AnnotationProcessingCompileTask.java:92)          at org.gradle.api.internal.tasks.compile.ResourceCleaningCompilationTask.call(ResourceCleaningCompilationTask.java:57)          at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:50)       at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:36)       at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.delegateAndHandleErrors(NormalizingJavaCompiler.java:100)      at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJa

**Comment by sjaakd** [comment]
@remipassmoilesel .. is this related to this issue? If not please open a new issue.. 

Q. Is the original issue solved? 

Looking at your log it could be a gradle issue.. Can you try with Maven?

**Comment by filiphr** [comment]
@remipassmoilesel looking at the stacktrace it seems like the jars are corrupt. Can you try to delete the mapstruct jars from your gradle and maven repo and then try again.

**Comment by remipassmoilesel** [comment]
Hi @filiphr, I have deleted my local repo and built mapstruct again but I have the same issue. I don't think this is an issue with Gradle as I use Mapstruct with Gradle very often.

**Comment by remipassmoilesel** [comment]
@sjaakd there is online a repo demonstrating the issue, with two simple tests, and this repo does not work so I think issue is not resolved.

 See https://github.com/remipassmoilesel/mapstruct-issue-03-2019

**Comment by filiphr** [comment]
I did the same @remipassmoilesel and I got the following:

```
> Task :compileJava FAILED
/Users/filiphr/projects_oss/mapstruct-issue-03-2019/src/main/java/org/remipassmoilesel/lightplayground/mapstruct/PersonMapper.java:16: warning: Unmapped target properties: "uuid, phoneNumber".
    PersonEntity toEntity(PersonDto dto);
                 ^
/Users/filiphr/projects_oss/mapstruct-issue-03-2019/src/main/java/org/remipassmoilesel/lightplayground/mapstruct/PersonMapper.java:21: error: No implementation type is registered for return type io.vavr.collection.List<org.remipassmoilesel.lightplayground.mapstruct.PersonEntity>.
    List<PersonEntity> toEntitiesVavr(List<PersonDto> dtos);
                       ^
1 error
1 warning
```

**Comment by filiphr** [comment]
Forgot to mention, this error is expected as MapStruct doesn't know how to handle Vavr Lists. If that method is removed or a default implementation is used then everything would compile and the test would be green. 

I just verified this locally

**Comment by remipassmoilesel** [comment]
You're right @filiphr, I tried once again and it work ! :+1: 

Thanks for your help and mapstruct, this is awesome !!!

**Comment by gerardbosch** [comment]
Hi there! Is there any chance this bugfix could be released prior 1.4.0? in any 1.3.x patch? I would really like to jump from 1.2.0.Final to 1.3.0 to take advantadge of some new features. But this issue stops us doing that. Thx!

P.S. In case it can't (and just for curiosity) is there any estimated time-frame for 1.4.0? Thx again!
